### ISSUE TITLE
Inference

### ISSUE THREAD

**Issue by SynixY** [zju3dv/MotionStreamer#24]
Hey team, can we get the inference script pls? And also the llama ckpt, i tried to full train the motion streamer using the Casual Tae babel t2m latents and my inference result was awful, here the code i did so far:
import torch
import numpy as np
import os
import smplx
from sentence_transformers import SentenceTransformer

# --- Model Imports ---
import models.tae as tae
from models.llama_model import LLaMAHF, LLaMAHFConfig

# --- Utility Imports ---
from utils import bvh, quat
from utils.face_z_align_util import rotation_6d_to_matrix, matrix_to_axis_angle, axis_angle_to_quaternion

# (The save_motion_as_bvh function remains unchanged)
def save_motion_as_bvh(motion_data_272, output_path, fps=60):
    # ... (code from your script)
    print(f"--- Starting Conversion to BVH: {os.path.basename(output_path)} ---")
    try:
        njoint = 22
        if isinstance(motion_data_272, torch.Tensor):
            motion_data_272 = motion_data_272.cpu().numpy()
        motion_data_272 = motion_data_272.squeeze(0)
        nfrm, _ = motion_data_272.shape
        rotations_6d = torch.from_numpy(motion_data_272[:, 8+6*njoint : 8+12*njoint]).reshape(nfrm, -1, 6)
        rotations_matrix = rotation_6d_to_matrix(rotations_6d).numpy()

        global_heading_diff_rot_6d = torch.from_numpy(motion_data_272[:, 2:8])
        global_heading_diff_rot = rotation_6d_to_matrix(global_heading_diff_rot_6d).numpy()

        velocities_root_xy = motion_data_272[:, :2]
        positions_no_heading = motion_data_272[:, 8 : 8+3*njoint].reshape(nfrm, -1, 3)
        height = positions_no_heading[:, 0, 1]
        global_heading_rot = [global_heading_diff_rot[0]]
        for R_rel in global_heading_diff_rot[1:]:
            global_heading_rot.append(np.matmul(R_rel, global_heading_rot[-1]))
        global_heading_rot = np.array(global_heading_rot)
        inv_global_heading_rot = np.transpose(global_heading_rot, (0, 2, 1))
        rotations_matrix[:, 0, ...] = np.matmul(inv_global_heading_rot, rotations_matrix[:, 0, ...])
        velocities_root_xyz = np.zeros((nfrm, 3))
        velocities_root_xyz[:, 0] = velocities_root_xy[:, 0]
        velocities_root_xyz[:, 2] = velocities_root_xy[:, 1]
        velocities_root_xyz[1:, :] = np.matmul(inv_global_heading_rot[:-1], velocities_root_xyz[1:, :, None]).squeeze(-1)
        root_translation = np.cumsum(velocities_root_xyz, axis=0)
        root_translation[:, 1] = height
        axis_angle = matrix_to_axis_angle(torch.from_numpy(rotations_matrix)).numpy().reshape(nfrm, -1)

        poses_24_joints = np.zeros((nfrm, 72))
        poses_24_joints[:, :66] = axis_angle
        model = smplx.create(model_path="body_models/human_model_files", model_type="smpl", gender="NEUTRAL")
        parents = model.parents.detach().cpu().numpy()

        rest_pose = model().joints.detach().cpu().numpy().squeeze()[:24,:]
        offsets = rest_pose - rest_pose[parents]
        offsets[0] = np.zeros(3)

        rotations_quat = axis_angle_to_quaternion(torch.from_numpy(poses_24_joints.reshape(-1, 24, 3))).numpy()
        rotations_euler = np.degrees(quat.to_euler(rotations_quat, order="zyx"))

        positions = np.zeros_like(rotations_quat[..., :3])
        positions[:, 0] = root_translation

        joint_names = [
            "Pelvis", "Left_hip", "Right_hip", "Spine1", "Left_knee", "Right_knee",
            "Spine2", "Left_ankle", "Right_ankle", "Spine3", "Left_foot", "Right_foot",
            "Neck", "Left_collar", "Right_collar", "Head", "Left_shoulder",
            "Right_shoulder", "Left_elbow", "Right_elbow", "Left_wrist", "Right_wrist",
            "Left_palm", "Right_palm",
        ]

        bvh.save(output_path, {
            "rotations": rotations_euler,
            "positions": positions,
            "offsets": offsets,
            "parents": parents,
            "names": joint_names,
            "order": "zyx",
            "frametime": 1.0 / fps,
        })
        print(f"✅ BVH file saved successfully to {output_path}")
    except Exception as e:
        print(f"❌ BVH Conversion Failed. Error: {e}")
        import traceback
        traceback.print_exc()

def main():
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    
    # --- CONFIGURATION ---
    history_motion_file = './babel_272/motion_data/001305.npy'
    mean_file = './babel_272/t2m_babel_mean_std/Mean.npy'
    std_file = './babel_272/t2m_babel_mean_std/Std.npy'
    causal_tae_checkpoint = './Causal_TAE_t2m_babel/net_last.pth'
    motionstreamer_checkpoint = './Experiments/motionstreamer_model/checkpoint_30000.pth'
    stop_token_file = 'reference_end_latent_t2m_272.npy'
    
    output_dir = "outputs"
    os.makedirs(output_dir, exist_ok=True)
    
    torch_dtype = torch.bfloat16
    print(f"Using data type: {torch_dtype}")
    
    num_steps_to_generate = 100 # How many new motion tokens (4-frame chunks) to create
    cfg_scale = 5.0
    temperature = 1.0
    text_prompt = "jump forward"

    # 1. --- Load All Models ---
    print("Loading models...")
    causal_tae = tae.Causal_HumanTAE(clip_range=[-30, 20])
    causal_tae.load_state_dict(torch.load(causal_tae_checkpoint, map_location='cpu')['net'])
    causal_tae.eval().to(device)

    t5_model = SentenceTransformer('sentencet5-xxl/').eval().to(device)
    for p in t5_model.parameters(): p.requires_grad = False
    
    config = LLaMAHFConfig.from_name('Normal_size')
    config.torch_dtype = "bfloat16" # Ensure model config matches training
    trans_encoder = LLaMAHF(config)
    trans_encoder.load_state_dict(torch.load(motionstreamer_checkpoint, map_location='cpu')['trans'])
    trans_encoder.eval().to(dtype=torch_dtype).to(device)
    print("✅ Models loaded.")

    # 2. --- Load Normalization Stats and Stop Token ---
    print("Loading normalization stats...")
    mean = torch.from_numpy(np.load(mean_file)).to(device)
    std = torch.from_numpy(np.load(std_file)).to(device)
    reference_end_token = torch.from_numpy(np.load(stop_token_file)).to(device, dtype=torch_dtype)
    print("✅ Stats loaded.")

    # 3. --- Prepare Initial Data for Streaming ---
    with torch.no_grad():
        # Encode initial history
        history_motion = torch.from_numpy(np.load(history_motion_file)).to(device).unsqueeze(0)
        save_motion_as_bvh(history_motion.clone(), os.path.join(output_dir, "history_motion.bvh"))
        normalized_history = (history_motion - mean) / std
        current_latents, _, _ = causal_tae.encode(normalized_history)
        current_latents = current_latents.unsqueeze(0).permute(0, 2, 1).to(dtype=torch_dtype)
        
        # Pre-calculate text embeddings once
        cond_text = torch.from_numpy(t5_model.encode([text_prompt])).to(device, dtype=torch_dtype)
        uncond_text = torch.from_numpy(t5_model.encode([""])).to(device, dtype=torch_dtype)

        print(f"--- Starting Streaming Generation with '{text_prompt}' ---")
        generated_latents_list = []
        max_context_len = 76 # Largest multiple of 4 less than 78-1

        for step in range(num_steps_to_generate):
            # Use the correct, dedicated function for single-token generation
            new_latent = trans_encoder.sample_for_eval_CFG_inference_next_one(
                current_token=[current_latents[:, -max_context_len:]], # Pass the recent history
                feat_clip_text=cond_text,
                empty_feat_clip_text=uncond_text,
                cfg=cfg_scale,
                temperature=temperature
            )
            
            # Append the new latent to our full history for the next step
            current_latents = torch.cat([current_latents, new_latent], dim=1)
            generated_latents_list.append(new_latent)

            # Check for stop condition
            distance = torch.sqrt(torch.sum((new_latent - reference_end_token)**2))
            print(f"Step {step+1}/{num_steps_to_generate} | Distance: {distance.item():.4f}")
            # Stricter threshold and a longer grace period before checking for the stop signal.
            # The original distance threshold of 3.0 was too high.
            if distance < 0.5 and step > 10:
                print("--- Stop condition met. ---")
                break
        
        # 4. --- Decode the final motion ---
        print("\n--- Decoding Final Motion ---")
        if not generated_latents_list:
            final_motion = history_motion
        else:
            newly_generated_latents = torch.cat(generated_latents_list, dim=1)
            new_motion_normalized = causal_tae.forward_decoder(newly_generated_latents.to(torch.float32))
            new_motion = new_motion_normalized * std.to(torch.float32) + mean.to(torch.float32)
            final_motion = torch.cat([history_motion, new_motion], dim=1)
            
        print(f"✅ Final motion shape: {final_motion.shape}")
        save_motion_as_bvh(final_motion, os.path.join(output_dir, "streamed_motion.bvh"))

if __name__ == "__main__":
    main()

But if you guys can provide the inference script used in the blender addon or in the demo of the page, and also the ckpt if you can, thanks anyways!
